includes/data-types/class-payment-information.php

```diff
@@ -0,0 +1,176 @@
+	 * @return \WC_Payment_Token/NULL The payment token.
+	 */
+	public function get_payment_token(): \WC_Payment_Token {
+		return $this->token;
+	}
+
+	/**
+	 * Update the payment token associated with this payment.
+	 *
+	 * @param \WC_Payment_Token $token The new payment token.
+	 */
+	public function set_token( \WC_Payment_Token $token ) {
+		$this->token = $token;
+	}
+
+	/**
+	 * Returns true if the payment token is not empty, false otherwise.
+	 *
+	 * @return bool True if payment token is not empty, false otherwise.
+	 */
+	public function is_using_saved_payment_method(): bool {
+		return ! empty( $this->token );
+	}
+
+	/**
+	 * Payment information constructor.
+	 *
+	 * @param array $request Associative array containing payment request information.
+	 * @param bool  $off_session Indicates whether the payment is merchant-initiated (true) or customer-initiated (false).
+	 *
+	 * @throws Exception - If no payment method is found in the provided request.
+	 */
+	public static function from_payment_request(
+		array $request,
+		bool $off_session = false
+	): Payment_Information {
+		$payment_method = self::get_payment_method_from_request( $request );
+		$token          = self::get_token_from_request( $request );
+		$off_session    = $off_session;
+
+		return new Payment_Information( $payment_method, $token, $off_session );
+	}
+
+	/**
+	 * Extracts the payment method from the provided request.
+	 *
+	 * @param array $request Associative array containing payment request information.
+	 *
+	 * @return string
+	 */
+	public static function get_payment_method_from_request( array $request ): string {
+		//phpcs:ignore WordPress.Security.ValidatedSanitizedInput.MissingUnslash
+		return ! empty( $request['wcpay-payment-method'] ) ? wc_clean( $request['wcpay-payment-method'] ) : '';
+	}
+
+	/**
+	 * Extract the payment token from the provided request.
+	 *
+	 * TODO: Once php requirement is bumped to >= 7.1.0 set return type to ?\WC_Payment_Token
+	 * since the return type is nullable, as per
+	 * https://www.php.net/manual/en/functions.returning-values.php#functions.returning-values.type-declaration
+	 *
+	 * @param array $request Associative array containing payment request information.
+	 *
+	 * @return \WC_Payment_Token|NULL
+	 */
+	public static function get_token_from_request( array $request ) {
+		if (
+			! isset( $request[ 'wc-' . \WC_Payment_Gateway_WCPay::GATEWAY_ID . '-payment-token' ] ) ||
+			'new' === $request[ 'wc-' . \WC_Payment_Gateway_WCPay::GATEWAY_ID . '-payment-token' ]
+		) {
+			return null;
+		}
+
+		//phpcs:ignore WordPress.Security.ValidatedSanitizedInput.MissingUnslash
+		$token = \WC_Payment_Tokens::get( wc_clean( $request[ 'wc-' . \WC_Payment_Gateway_WCPay::GATEWAY_ID . '-payment-token' ] ) );
+
+		// If the token doesn't belong to this gateway or the current user it's invalid.
+		if ( ! $token || \WC_Payment_Gateway_WCPay::GATEWAY_ID !== $token->get_gateway_id() || $token->get_user_id() !== get_current_user_id() ) {
```

Comment:
We might need to check if `$token->get_user_id() !== get_current_user_id()` needs to be changed in order to work for scheduled renewals or renewals triggered by the merchant.

Reply:
I did a quick test and it seems like there's no need to change this. The only place where this gets called is in `process_payment` and I believe we need to check if the logged-in user can use the token at that time. For subscription renewals, we manually create `Payment_Information` with the token saved in the order.